### PR TITLE
NonceHandlerV3

### PR DESCRIPTION
**Pull request by cristure**
I reworked the handler a bit. The main addititon in the `transactionWorker` similiar to how the resendTransaction goroutine was working in the previous iteration. Now all the transactions are being processed by this worker which will retrieve the transaction from the `transactionQueue` which unde the hood is a heap prioritizing over the minimum nonce. That means that everytime we want to retrieve a transaction it will return the one with the lowest nonce.

The workflow is prety straightforward. Everytime an `adressNonceHandler` is created, such a goroutine `transactionWorker.start()` that will poll the `transactionQueue` will start. When  `SendTransaction` is called, the transaction passed in the arguments will end in the queue and futher processed by the `transactionWorker` and futher send the road.

By doing this we make sure to always send transactions in order, there are some caveats to it that will need to be addressed though.

### CODE REVIEW

File: interactors/nonceHandlerV3/workers/transactionWorker.go [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+	return tx
+}
+
+// Len required by sort.Interface
+func (tq *transactionQueue) Len() int {
+	return len(*tq)
+}
+
+// Swap required by the sort.Interface
+func (tq transactionQueue) Swap(a, b int) {
```
Comment (iulianpascalau) [inline interactors/nonceHandlerV3/workers/transactionWorker.go:49]: from linter: 
> Struct transactionQueue has methods on both value and pointer receivers. Such usage is not recommended by the Go Documentation.

File: interactors/nonceHandlerV3/workers/transactionWorker.go [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+
+	heap.Push(&tw.tq, transaction)
+	return r
+}
+
+// start will spawn a goroutine tasked with iterating all the transactions inside the priority queue. The priority is
+// given by the nonce, meaning that transaction with lower nonce will be sent first.
+func (tw *TransactionWorker) start(pollingInterval time.Duration) {
+	go func() {
+		for {
```
Comment (iulianpascalau) [inline interactors/nonceHandlerV3/workers/transactionWorker.go:100]: resource leakage: this go routine never ends

File: interactors/nonceHandlerV3/workers/transactionWorker.go [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+			// Retrieve the transaction in the queue with the lowest nonce.
+			tx := tw.nextTransaction()
+
+			// If there are no transaction in the queue, the result will be nil.
+			// That means there are no transactions to send.
+			if tx == nil {
+				time.Sleep(pollingInterval)
```
Comment (iulianpascalau) [inline interactors/nonceHandlerV3/workers/transactionWorker.go:107]: suggestion to switch to a select construction here to use the provided context. Also, use a ticker.

File: interactors/nonceHandlerV3/workers/transactionWorker.go [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+func (tq *transactionQueue) Pop() interface{} {
+	currentQueue := *tq
+	n := len(currentQueue)
+	tx := currentQueue[n-1]
+	*tq = currentQueue[0 : n-1]
+	return tx
+}
+
+// Len required by sort.Interface
+func (tq *transactionQueue) Len() int {
```
Comment (iulianpascalau) [inline interactors/nonceHandlerV3/workers/transactionWorker.go:44]: where is the slice.Sort called?

File: interactors/nonceHandlerV3/workers/transactionWorker_test.go [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+	// We add roughly at the same time un-ordered transactions.
+	w.AddTransaction(&transaction.FrontendTransaction{Nonce: 91})
+	w.AddTransaction(&transaction.FrontendTransaction{Nonce: 1})
+	w.AddTransaction(&transaction.FrontendTransaction{Nonce: 13})
+	w.AddTransaction(&transaction.FrontendTransaction{Nonce: 10})
+	w.AddTransaction(&transaction.FrontendTransaction{Nonce: 99})
+	w.AddTransaction(&transaction.FrontendTransaction{Nonce: 8})
+	w.AddTransaction(&transaction.FrontendTransaction{Nonce: 7})
+
+	// Verify that the results come in ordered.
```
Comment (iulianpascalau) [inline interactors/nonceHandlerV3/workers/transactionWorker_test.go:35]: Does this really test that the sent transactions were sent in order?

**Review by iulianpascalau — CHANGES_REQUESTED** [review]

**Review by iulianpascalau — DISMISSED** [review]

File: interactors/nonceHandlerV3/addressNonceHandler.go [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+// Close will cancel all related processes..
+func (anh *addressNonceHandler) Close() {
+	anh.cancelFunc()
+}
+
+func (anh *addressNonceHandler) fetchGasPriceIfRequired(ctx context.Context) {
+	if anh.gasPrice == 0 {
+		networkConfig, err := anh.proxy.GetNetworkConfig(ctx)
+
+		anh.mut.Lock()
```
[the remark this answers is not in the document]
Comment (cristure) [reply interactors/nonceHandlerV3/addressNonceHandler.go:99]: changed. good catch

File: interactors/nonceHandlerV3/nonceTransactionsHandler.go [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+		}
+
+		idx := i
+		g.Go(func() error {
+			sentHash, errSend := anh.SendTransaction(ctx, &txCopy)
+			if errSend != nil {
+				return fmt.Errorf("%w while sending transaction for address %s", errSend, addrAsBech32)
+			}
+
+			sentHashes[idx] = sentHash
```
[the remark this answers is not in the document]
Comment (cristure) [reply interactors/nonceHandlerV3/nonceTransactionsHandler.go:173]: I believe you can write concurrently on a slice as long as you are providing an index, since you are not writing to the same variable, nor changing the slice header inside the `errgroup`. Further reads on the slice cannot be made until all goroutines are finished, hence the `g.Wait()` at the end. I've also tested this function with the -race condition and so far it hasn't returned any issues.

File: interactors/nonceHandlerV3/nonceTransactionsHandler.go [hunk 8]
```diff
@@ -48,13 +49,13 @@ func NewNonceTransactionHandlerV3(args ArgsNonceTransactionsHandlerV3) (*nonceTr
 		return nil, interactors.ErrNilProxy
 	}
 	if args.PollingInterval < minimumIntervalToResend {
-		return nil, fmt.Errorf("%w for pollingInterval in NewNonceTransactionHandlerV2", interactors.ErrInvalidValue)
+		return nil, fmt.Errorf("%w for intervalToSend in NewNonceTransactionHandlerV2", interactors.ErrInvalidValue)
```
[the remark this answers is not in the document]
Comment (cristure) [reply interactors/nonceHandlerV3/nonceTransactionsHandler.go:52]: right. missed that

File: interactors/nonceHandlerV3/addressNonceHandler.go [hunk 9]
```diff
@@ -55,18 +56,19 @@ func NewAddressNonceHandlerV3(proxy interactors.Proxy, address sdkCore.AddressHa
 // ApplyNonceAndGasPrice will apply the computed nonce to the given FrontendTransaction
 func (anh *addressNonceHandler) ApplyNonceAndGasPrice(ctx context.Context, txs ...*transaction.FrontendTransaction) error {
```
[the remark this answers is not in the document]
Comment (cristure) [reply interactors/nonceHandlerV3/addressNonceHandler.go:57]: no, you didn't miss anything. I didn't think of applying the nonce and gas price asyncronously, but I think it'd make some sense to do it. Though, this will need some refactoring and rethinking.

File: interactors/nonceHandlerV3/addressNonceHandler.go [hunk 10]
```diff
@@ -54,19 +57,20 @@ func NewAddressNonceHandlerV3(proxy interactors.Proxy, address sdkCore.AddressHa
 
 // ApplyNonceAndGasPrice will apply the computed nonce to the given FrontendTransaction
 func (anh *addressNonceHandler) ApplyNonceAndGasPrice(ctx context.Context, txs ...*transaction.FrontendTransaction) error {
-	nonce, err := anh.fetchNonce(ctx)
-	for i, tx := range txs {
-		tx.Nonce = nonce + uint64(i)
+	for _, tx := range txs {
+		//anh.mut.Lock()
```
Comment (iulianpascalau) [inline interactors/nonceHandlerV3/addressNonceHandler.go:61]: why the commented code with lock/unlock?

File: interactors/nonceHandlerV3/addressNonceHandler.go [hunk 11]
```diff
@@ -75,6 +79,18 @@ func (anh *addressNonceHandler) SendTransaction(ctx context.Context, tx *transac
 
 	select {
 	case response := <-ch:
+		if response.Error == nil {
```
Comment (iulianpascalau) [inline interactors/nonceHandlerV3/addressNonceHandler.go:82]: can be extracted in a new function, no need for the else branch

File: interactors/nonceHandlerV3/addressNonceHandler.go [hunk 12]
```diff
@@ -92,26 +108,37 @@ func (anh *addressNonceHandler) Close() {
 	anh.cancelFunc()
 }
 
-func (anh *addressNonceHandler) fetchGasPriceIfRequired(ctx context.Context) {
+func (anh *addressNonceHandler) applyGasPriceIfRequired(ctx context.Context, tx *transaction.FrontendTransaction) {
+	anh.mut.Lock()
```
Comment (iulianpascalau) [inline interactors/nonceHandlerV3/addressNonceHandler.go:112]: why this change? This will keep the lock as long at the API get call is done.

File: interactors/nonceHandlerV3/addressNonceHandler.go [hunk 13]
```diff
@@ -92,26 +108,37 @@ func (anh *addressNonceHandler) Close() {
...
 
-func (anh *addressNonceHandler) fetchNonce(ctx context.Context) (uint64, error) {
-	account, err := anh.proxy.GetAccount(ctx, anh.address)
-	if err != nil {
-		return 0, err
-	}
+func (anh *addressNonceHandler) fetchNonce(ctx context.Context) (int64, error) {
+	// if it is the first time applying nonces to this address, or if the cache was invalidated it will try to fetch
+	// the nonce from the chain.
+	anh.mut.Lock()
```
Comment (iulianpascalau) [inline interactors/nonceHandlerV3/addressNonceHandler.go:131]: also here. The mutex is kept longer than necessary.

File: interactors/nonceHandlerV3/nonceTransactionsHandler_test.go [hunk 14]
```diff
@@ -231,6 +233,173 @@ func TestSendTransactionsCloseDelay(t *testing.T) {
...
+			Value:    "5000000000000000000",
+			Nonce:    uint64(i),
+			GasPrice: 1000000000,
+			Version:  2,
+		}
+		txs = append(txs, tx)
+	}
+
+	var wg sync.WaitGroup
+	wg.Add(1)
```
Comment (iulianpascalau) [inline interactors/nonceHandlerV3/nonceTransactionsHandler_test.go:261]: L261-L304 can be done in a for loop.
Also, no need to panic the code in the go routine, can use a require.NotNil for the inner err

File: interactors/nonceHandlerV3/nonceTransactionsHandler_test.go [hunk 15]
```diff
@@ -231,6 +233,173 @@ func TestSendTransactionsCloseDelay(t *testing.T) {
...
+		Nonce:    0,
+		GasPrice: 1000000000,
+		Version:  2,
+	}
+
+	wg := sync.WaitGroup{}
+	var errCount int
+	var sentCount int
+
+	wg.Add(1)
```
Comment (iulianpascalau) [inline interactors/nonceHandlerV3/nonceTransactionsHandler_test.go:341]: for loop fo L341-L365?

File: interactors/nonceHandlerV3/addressNonceHandler.go [hunk 16]
```diff
@@ -92,26 +108,37 @@ func (anh *addressNonceHandler) Close() {
...
 	}
+	tx.GasPrice = core.MaxUint64(anh.gasPrice, tx.GasPrice)
 }
 
-func (anh *addressNonceHandler) fetchNonce(ctx context.Context) (uint64, error) {
-	account, err := anh.proxy.GetAccount(ctx, anh.address)
-	if err != nil {
-		return 0, err
-	}
+func (anh *addressNonceHandler) fetchNonce(ctx context.Context) (int64, error) {
```
Comment (iulianpascalau) [inline interactors/nonceHandlerV3/addressNonceHandler.go:128]: this is now "computeNonce" because the fetch is not done each time

File: interactors/nonceHandlerV3/workers/transactionWorker.go [hunk 17]
```diff
@@ -104,35 +105,39 @@ func (tw *TransactionWorker) AddTransaction(transaction *transaction.FrontendTra
 		return r
 	}
 
-	tw.responsesChannels[transaction.Nonce] = r
+	// check if a tx with the same nonce is currently being sent.
+	if _, ok := tw.responsesChannels[transaction.Nonce]; ok {
```
Comment (iulianpascalau) [inline interactors/nonceHandlerV3/workers/transactionWorker.go:109]: :+1: very good catch here

File: interactors/nonceHandlerV3/workers/transactionWorker.go [hunk 18]
```diff
@@ -104,35 +105,39 @@ func (tw *TransactionWorker) AddTransaction(transaction *transaction.FrontendTra
...
 			select {
 			case <-ctx.Done():
 				log.Info("context cancelled - transaction worker has stopped")
 				tw.closeAllChannels(ctx)
 				return
 			case <-ticker.C:
-				tw.processNextTransaction(ctx, chMoreWorkToDo)
-			case <-chMoreWorkToDo:
-				tw.processNextTransaction(ctx, chMoreWorkToDo)
+				tw.processNextTransaction(ctx)
```
Comment (iulianpascalau) [inline interactors/nonceHandlerV3/workers/transactionWorker.go:134]: Is this change desired? Now the component will send one transaction at the interval time. In the previous variant it would have sent pending transactions one after another

**Review by iulianpascalau — CHANGES_REQUESTED** [review]

File: interactors/nonceHandlerV3/workers/transactionWorker.go (hunk 18 above)
Comment (cristure) [reply interactors/nonceHandlerV3/workers/transactionWorker.go:134]: the proxy can't handle that many requests anyway. This way you can leverage the `intervalToSend` in order to be sure that 100% of the transactions sent will be executed.

File: interactors/nonceHandlerV3/addressNonceHandler.go (hunk 10 above)
Comment (cristure) [reply interactors/nonceHandlerV3/addressNonceHandler.go:61]: forgot about it. thanks :+1:

File: interactors/nonceHandlerV3/addressNonceHandler.go (hunk 11 above)
Comment (cristure) [reply interactors/nonceHandlerV3/addressNonceHandler.go:82]: done.

File: interactors/nonceHandlerV3/addressNonceHandler.go (hunk 12 above)
Comment (cristure) [reply interactors/nonceHandlerV3/addressNonceHandler.go:112]: changed.

File: interactors/nonceHandlerV3/addressNonceHandler.go (hunk 13 above)
Comment (cristure) [reply interactors/nonceHandlerV3/addressNonceHandler.go:131]: changed.

File: interactors/nonceHandlerV3/nonceTransactionsHandler_test.go (hunk 14 above)
Comment (cristure) [reply interactors/nonceHandlerV3/nonceTransactionsHandler_test.go:261]: done

File: interactors/nonceHandlerV3/addressNonceHandler.go [hunk 19]
```diff
@@ -58,14 +59,12 @@ func NewAddressNonceHandlerV3(proxy interactors.Proxy, address sdkCore.AddressHa
 // ApplyNonceAndGasPrice will apply the computed nonce to the given FrontendTransaction
 func (anh *addressNonceHandler) ApplyNonceAndGasPrice(ctx context.Context, txs ...*transaction.FrontendTransaction) error {
 	for _, tx := range txs {
-		//anh.mut.Lock()
-		nonce, err := anh.fetchNonce(ctx)
+		nonce, err := anh.computeNonce(ctx)
 		if err != nil {
 			//anh.mut.Unlock()
```
Comment (iulianpascalau) [inline interactors/nonceHandlerV3/addressNonceHandler.go:64]: comment remained here

File: interactors/nonceHandlerV3/addressNonceHandler.go [hunk 20]
```diff
@@ -109,23 +111,28 @@ func (anh *addressNonceHandler) Close() {
 }
 
 func (anh *addressNonceHandler) applyGasPriceIfRequired(ctx context.Context, tx *transaction.FrontendTransaction) {
+	var (
+		networkConfig *data.NetworkConfig
+		err           error
+	)
 	anh.mut.Lock()
```
Comment (iulianpascalau) [inline interactors/nonceHandlerV3/addressNonceHandler.go:118]: RLock/RUnlock?

File: interactors/nonceHandlerV3/addressNonceHandler.go (hunk 13 above)
Comment (iulianpascalau) [reply interactors/nonceHandlerV3/addressNonceHandler.go:138]: yup, true. I guess this should remain like this as it only keeps the lock only if the fetch is needed. Which should block any other access as to have consistent values.

File: interactors/nonceHandlerV3/workers/transactionWorker.go (hunk 18 above)
Comment (iulianpascalau) [reply interactors/nonceHandlerV3/workers/transactionWorker.go:134]: We can not make assumptions that the proxy can not handle a lot of requests. The public proxy might have some antiflood protections but that might not be the case on some custom setups where this service might be installed. You can rework this code in 2 different ways:
1. you can add a different ticker with a different interval for consecutive sends, or
2. can use the proxy.SendTransactions to send a batch of transactions.

"Putting delays" usually solves some problems but degrades the performance

Comment (cristure) [reply interactors/nonceHandlerV3/workers/transactionWorker.go:134]: I could decrease the minimum limit. One could put 1 millisecond interval  instead and still retain flexibility.

File: interactors/nonceHandlerV3/addressNonceHandler.go (hunk 19 above)
Comment (cristure) [reply interactors/nonceHandlerV3/addressNonceHandler.go:64]: I am pretty sure it's not there any longer. Might be github's fault.

File: interactors/nonceHandlerV3/addressNonceHandler.go (hunk 20 above)
Comment (cristure) [reply interactors/nonceHandlerV3/addressNonceHandler.go:118]: changed.

**Review by iulianpascalau — APPROVED** [review]

File: interactors/nonceHandlerV3/nonceTransactionsHandler.go [hunk 21]
```diff
@@ -0,0 +1,195 @@
...
+	err := g.Wait()
+
+	return sentHashes, err
+}
+
+// Close will cancel all related processes.
+func (nth *nonceTransactionsHandlerV3) Close() {
+	nth.mutHandlers.RLock()
+	defer nth.mutHandlers.RUnlock()
+	for _, handler := range nth.handlers {
```
Comment (AdoAdoAdo) [inline interactors/nonceHandlerV3/nonceTransactionsHandler.go:187]: could get under mutex a copy of the handlers map and then do the close without any held mutex.

I see that close could take more time due to the possibility of interacting with the proxy in certain conditions.

**Review by AdoAdoAdo — APPROVED** [review]

File: interactors/nonceHandlerV3/nonceTransactionsHandler.go [hunk 22]
```diff
@@ -183,8 +183,10 @@ func (nth *nonceTransactionsHandlerV3) SendTransactions(ctx context.Context, txs
 // Close will cancel all related processes.
 func (nth *nonceTransactionsHandlerV3) Close() {
 	nth.mutHandlers.RLock()
-	defer nth.mutHandlers.RUnlock()
-	for _, handler := range nth.handlers {
+	handlers := nth.handlers
```
Comment (iulianpascalau) [inline interactors/nonceHandlerV3/nonceTransactionsHandler.go:186]: I do not think this is correct as the slice pointer remains the same. Please use `copy` here

**Review by AdoAdoAdo — APPROVED** [review]

**Review by iulianpascalau — APPROVED** [review]
